### PR TITLE
Refactoring Authentication/Sign Up process

### LINKED ISSUE #60: Refactor Authentication / Sign up workflow
All users should be able to log in immediately and then sign up to be a reviewer or reviewee from there. This allows reviewees to be logged in and submit the review form multiple times if they wish to submit a new resume / request a new review and be rematched. 

Also, users can only comment on their My Account page once they are matched.

### PR DESCRIPTION
**Pull request by shreyabarua**
Refactored all my code from last week. Here's the new workflow:
1. All users can immediately Log In to ResumeBuddy. This registers them as a user in the User DB and only holds their email.
2. Once logged in, users can fill up Sign Up forms and their User entity will update with new info (from Ani's PR).
3. Once matched, all users can successfully see their own as well as their match's comments on the My Account page (this is done by every match being assigned a unique ID) --> has been tested on newly deployed website

I know the branch is called Local Storage but I ended up not using it as the unique Match ID fixed the issue of overlapping data when multiple people were logged in at the same time.

### CODE REVIEW

File: resume-buddy/src/main/java/com/google/sps/data/Comment.java [hunk 1]
```diff
@@ -48,8 +48,11 @@ public String getId() {
...
-    @Override
-    public int compare(Comment a, Comment b) { return (b.getDate()).compareTo(a.getDate()); }
-  };
+  public static final Comparator<Comment> ORDER_BY_DATE =
+      new Comparator<Comment>() {
+        @Override
+        public int compare(Comment a, Comment b) {
+          return (b.getDate()).compareTo(a.getDate());
+        }
+      };
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/data/Comment.java]: Hm, was this auto-formatted again? If you can change it back to what it was before, that'd be great.

File: resume-buddy/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 2]
```diff
@@ -79,72 +66,31 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
-    query.setFilter(uuidPropertyFilter);
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    PreparedQuery results = datastore.prepare(query);
-
-    return results.countEntities(FetchOptions.Builder.withDefaults()) != 0;
-  }
-
-  /* Checks if the given user of the given userType has a match of the opposite userType. */
-  public static boolean hasMatch(UserType userType, String email) {
+  public static String getMatchID(String email) {
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/servlets/CommentServlet.java:78]: Add a comment above the function explaining what it does.

File: resume-buddy/src/main/java/com/google/sps/servlets/CommentServlet.java [hunk 3]
```diff
@@ -79,72 +66,31 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
-      } else {
-        return (String) entity.getProperty("reviewee");
-      }
-    }
-    return "";
-  }
-
-  /* add comments from given email and userType(reviewer or reviewee) */
-  public static void addComments(UserType userType, String email, List<Comment> comments) {
+  /* add all comments that have the given match id */
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/servlets/CommentServlet.java]: Nit: I would change this comment to say "Populates the given list of Comments with all comments that have the given match ID."

File: resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 4]
```diff
@@ -35,29 +34,33 @@
...
-  /* TODO: Find a way for the user type to be passed to the CommentsServlet from JavaScript
-   * (instead of being stored here as a static variable) */
-  private static UserType userType;
-
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
     UserService userService = UserServiceFactory.getUserService();
-    boolean isValidUser;
+    boolean status = userService.isUserLoggedIn();
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java]: Nit: You don't need this variable, just use `userService.isUserLoggedIn()` for readability.

File: resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 5]
```diff
@@ -67,29 +70,14 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    Query query;
-    if (userType == UserType.REVIEWEE) {
-      query = new Query("Reviewee");
-    } else {
-      query = new Query("Reviewer");
-    }
-    Filter emailFilter = new FilterPredicate("email", FilterOperator.EQUAL, email_key);
+  /* return true if this user does not exist in the User db yet*/
+  public static boolean newUser(String email) {
```
Comment (oydang) [inline resume-buddy/src/main/java/com/google/sps/servlets/LoginServlet.java]: Nit: rename to `isNewUser`

File: resume-buddy/src/main/webapp/scripts/resume-review-script.js [hunk 6]
```diff
@@ -7,6 +7,9 @@ function onLoad() {
  * Fetches comments from the servers and adds them to the DOM.
  */
 function getComments() {
+  /* TODO: default hide comments section
+  /* Only if the user has a match, display comments functionality
+   */
```
Comment (oydang) [inline resume-buddy/src/main/webapp/scripts/resume-review-script.js]: Please make a corresponding GitHub Issue for this TODO and link to it in this comment.

File: resume-buddy/src/test/java/com/google/sps/CommentsTest.java [hunk 7]
```diff
@@ -34,46 +35,38 @@
...
-  }
-
-  @Test
-  public void getMatch2() throws ParseException {
-    datastore.put(match);
-    String matchUser = CommentServlet.getMatch(UserType.REVIEWER, "sundar@google.com");
-    Assert.assertEquals("shreyabarua@google.com", matchUser);
-  }
-
+  // TODO: fix this test!
```
Comment (oydang) [inline resume-buddy/src/test/java/com/google/sps/CommentsTest.java]: Please add a GitHub Issue for this TODO and link it in the comment.

**Review by oydang — CHANGES_REQUESTED** [review]
Great job! This is a significant refactor.

File: resume-buddy/src/test/java/com/google/sps/CommentsTest.java (hunk 7 above)
Comment (shreyabarua) [reply resume-buddy/src/test/java/com/google/sps/CommentsTest.java]: Sorry this test was already fixed, I just forgot to remove the TODO. Got rid of it now!

**Review by oydang — APPROVED** [review]

**Comment by AniMachaidze** [comment]
Closes #60

**Pull request merged**